Consistently check for liveness of endpoint when re-locking

nebril opened the pull request:
fixes #4997

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5116)
<!-- Reviewable:end -->

nebril committed:
Hide pkg/endpoint/Endpoint.Mutex behind wrappers

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2608,3 +2659,71 @@ func (e *Endpoint) syncPolicyMapController() {
+	if e.IsDisconnecting() {
+		return fmt.Errorf("Endpoint was removed before lock attempt.")
+	}
+	e.mutex.Lock()
```
I think that this should occur before the `e.IsDisconnecting()` call accesses `e.state` without holding a lock.

nebril replied:
Can you point me to this call? This method should not be called without error handling, which should abort all accesses, since lock is not acquired.

nebril replied:
@joestringer ^^

joestringer replied:
The comment above `e.IsDisconnecting()` describes how it must be called while holding `e.mutex.Lock()` to ensure safe access; otherwise there is no guarantee that by the time the call to `IsDisconnecting()` returns that the state hasn't changed.

`IsDisconnecting()` itself reads `e.state`, which should be protected by the endpoint Lock.

In this new Lock() call, `e.state` is being checked without grabbing the lock. So, it seems plausible that this branch statement is checked from one goroutine, then some other code updates the state, then this code grabs the lock. In that race condition, `e.Lock()` will return success with a locked endpoint in disconnecting state.

nebril replied:
Ah, thanks, I missed this.

nebril replied:
Done

nebril committed:
Change endpoint mutex usage in daemon/*

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

joestringer reviewed (commented):
Minor nit, other parts of the codebase (and in general in golang) use a `Lock()` that is guaranteed to lock, ie `func Lock()`.

This PR currently changes this signature for part of the codebase, then introduces `UnconditionalLock()`.

It may be more idiomatic to instead introduce a new `func TryLock() error` signature for the version that may fail, then keep `Lock()` as a guaranteed lock.

nebril committed:
Fix mutex calls in unit tests

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2622,7 +2678,80 @@ func (e *Endpoint) syncPolicyMapController() {
+
+// RUnlock read unlocks endpoint mutex and returns error if endpoint was removed
+func (e *Endpoint) RUnlockAlive() error {
+	if e.IsDisconnecting() {
```
This is not necessary. The endpoint cannot transition its state while it is being locked.

nebril replied:
done

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -514,8 +526,16 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, boo
+			lockerr = e.LockAlive()
+			if lockerr != nil {
+				// holding the mutex to log the error
+				e.UnconditionalLock()
```
Can be an RLock

nebril replied:
done

nebril committed:
Simplify endpoint unlock

Endpoint unlocking doesn't need endpoint state check since locked
endpoint cannot change its state

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Adjust endpoint logging lock

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Adjust daemon unit tests to new ep locking

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

tgraf commented on daemon/endpoint.go:
```diff
@@ -429,7 +425,11 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 	}
-	ep.Mutex.Unlock()
+
+	// If endpoint is disconnected, the request should fail
```
This seems wrong, we should be unlocking here.

nebril replied:
done

tgraf commented on daemon/state.go:
```diff
@@ -97,12 +97,16 @@ func (d *Daemon) restoreOldEndpoints(dir string, clean bool) (*endpointRestoreSt
 		}
 
-		ep.Mutex.Lock()
+		if lockerr := ep.LockAlive(); lockerr != nil {
```
This can be an unconditional lock. We are not serving the API yet and using `LockAlive()` suggests that the endpoint to be restored could be manipulated.

nebril replied:
done

tgraf commented on daemon/state.go:
```diff
@@ -144,28 +148,32 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
 		// endpointmanager.
-		ep.Mutex.RLock()
+
+		if lockerr := ep.RLockAlive(); lockerr != nil {
```
This can be an unconditional lock as the endpoint has not been inserted into the endpointmanager yet and cannot be manipulated.

nebril replied:
done

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -462,14 +465,15 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, boo
 	buildStart := time.Now()
 
-	e.Mutex.Lock()
+	if err = e.LockAlive(); err != nil {
```
This will help massively.

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -514,8 +519,16 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, boo
 	defer func() {
-		if err != nil {
-			e.Mutex.Lock()
+		if reterr != nil {
```
Where is `reterr` coming from?

nebril replied:
reterr is named return value defined at the beginning of method definition on line 454, it can be manipulated in deferred function calls.

tgraf replied:
Why is this change to `reterr` required?

nebril replied:
Good point, after change to UnconditionalLock - it's not required. Will change

EDIT:
On second thought, I think that it's cleaner if `reterr` is checked here, since there can be other error returned from this method which isn't assigned to `err` variable.

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -514,8 +519,16 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, boo
-		if err != nil {
-			e.Mutex.Lock()
+		if reterr != nil {
+			lockerr = e.LockAlive()
```
This must be unconditional as we otherwise leak resources that have already created.

nebril replied:
done

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -635,39 +648,42 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, boo
 		e.bpfHeaderfileHash = bpfHeaderfilesHash
 	} else {
-		e.Mutex.RLock()
+		if lockerr = e.RLockAlive(); lockerr != nil {
```
Let's keep it unconditional if locking is only required for logging like this.

nebril replied:
done

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -1592,19 +1576,23 @@ func (e UpdateStateChangeError) Error() string { return e.msg }
 // or if endpoint regeneration was unable to be triggered.
 func (e *Endpoint) Update(owner Owner, cfg *models.EndpointConfigurationSpec) error {
-	e.Mutex.Lock()
+	var lockerr error
```
Why do use `lockerr` instead of the standard `err` convention?

nebril replied:
I started using it in bpf.go because `err` was used in several big functions in a convoluted way and after that I decided that it's nice to have a distinctly named error just for lock errors. But maybe it's too much. Can change it.

tgraf replied:
Please change it. We will likely backport this and it should be as small as possible.

nebril replied:
Done

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2359,8 +2363,8 @@ func (e *Endpoint) identityLabelsChanged(owner Owner, myChangeRev int) error {
 	if e.identityResolutionIsObsolete(myChangeRev) {
-		e.Mutex.RUnlock()
-		elog.Debug("Endpoint identity has changed, aborting resolution routine in favour of new one")
+		defer elog.Debug("Endpoint identity has changed, aborting resolution routine in favour of new one")
```
Why the defer? I think it's fine to move it out of the critical section but then let's just swap and not sure defer.

nebril replied:
Seems like a leftover from when RUnlock returned an error. Done

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2606,9 +2613,11 @@ func (e *Endpoint) syncPolicyMapController() {
-				e.Mutex.Lock()
-				defer e.Mutex.Unlock()
+			DoFunc: func() (reterr error) {
+				if lockerr := e.LockAlive(); lockerr != nil {
```
This will help a lot as well! I don't think we have a return an error here because the endpoint disappearing means the regular end of the controller.

tgraf replied:
We can add a debug message if we want but it's not an error condition.

nebril replied:
Added a debug message and returning no error here then.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2622,7 +2631,59 @@ func (e *Endpoint) syncPolicyMapController() {
+	e.mutex.Lock()
+	if e.IsDisconnecting() {
+		e.mutex.Unlock()
+		return fmt.Errorf("endpoint was removed before lock attempt")
```
`lock failed: endpoint is in the process of being removed`

nebril replied:
done

nebril committed:
Fix endpoint locking review issues

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -708,13 +711,20 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
+		if lockerr := e.LockAlive(); lockerr != nil {
+			if retErr == nil {
+				retErr = lockerr
+			} else {
```
I don't understand the else branch here. When is retErr ever not nil? Is this trying to protect when we add another defer before this that has already set `retErr`?

nebril replied:
`retErr` is not nil if the `regenerate` method returns any non-nil `error`. So if there is no error returned from `regenerate`, but lock from line 714 fails - we will return the lock fail error. If there already is an error stored in `retErr` (meaning there was a regeneration error) - we return this and log the locking error.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -884,8 +887,9 @@ func (e *Endpoint) GetModel() *models.Endpoint {
 	}
-	e.Mutex.RLock()
-	defer e.Mutex.RUnlock()
+	// NOTE: Using rlock on mutex directly because GetModelRLocked handles endpoint in disco state properly
```
spelling

nebril replied:
what is wrong with spelling here?

tgraf replied:
I'm all for partying but `disco` is not a state ;-)

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2102,24 +2095,23 @@ OKState:
-// policy revision as setup by e.regenerate(). endpoint.Mutex should not be held.
-func (e *Endpoint) bumpPolicyRevision(revision uint64) {
-	e.Mutex.Lock()
+// policy revision as setup by e.regenerate(). endpoint.Mutex should held.
```
spelling

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2295,11 +2291,13 @@ func (e *Endpoint) UpdateLabels(owner Owner, identityLabels, infoLabels pkgLabel
 
-	e.Mutex.Lock()
+	// NOTE: not sure if this should be allowed, please take a look in review
+	e.UnconditionalLock()
```
This should check for liveness

nebril replied:
Should all UpdateLabels calls check for lock error? Should docker and cri `handleCreateWorkload` fail when this lock fails?

tgraf replied:
This lock should fail as it will prevent changing the labels and triggering an identity resolution for an endpoint that is already being deleted.

tgraf replied:
Let's review and fid the container runtime handlers separately outside of the scope of this PR. At least CRI is currently missing to hold ep.Mutex while calling `endpointmanager.UpdateReferences()` so it might potentially create a entry for an endpoint that is being deleted. Let's just make sure that we don't update the labels of the endpoint for now.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -2323,10 +2318,14 @@ func (e *Endpoint) identityResolutionIsObsolete(myChangeRev int) bool {
+	// NOTE: need to figure out if we need e.LockAlive() here
+	// The disco error will be reported by resolve-identity-e.ID controller
+	// which seems to be enough
+	e.UnconditionalLock()
```
Unconditional lock is fine here, this is only logging. The main purpose for checking the liveness is to prevent bigger blocks of work that are about to be scheduled.

aanm reviewed (changes requested):
This is a my personal POV. I believe this will drastically create developer errors as we are introducing more functions that can be used to lock and unlock an endpoint. `RLockAlive`, `UnconditionalRLock`, `RUnlock`, `Unlock`, `Lock`? Did I miss one more?

I thought the whole purpose of the state machine was to avoid the bug this PR is fixing, the state machine should be the one that returns errors, IMHO a developer shouldn't care if a "humble" `Lock` returns an error or not.

If the state machine main purpose is not doing what we want it to do, then we should fix that and not introduce more complexity into the code.

@nebril do not consider this an attack to your code as this is a merely observation on my side over the overall code complexity that we are all aware of.

nebril committed:
Clean up various typos

Signed-off-by: Maciej Kwiek <maciej@covalent.io>
